File: frontend/src/stores/zabbixTemplates.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { apiService } from '@/services/apiService'
import type { ZabbixTemplateItemSummary } from './devices'

export interface ZabbixTemplate {
  id: number
  zabbixUuid: string | null
  name: string
  description: string | null
  zabbixVersion: string | null
  importedAt: string
  deviceCount: number
  items: ZabbixTemplateItemSummary[]
}

export interface ZabbixImportResult {
  id: number
  name: string
  itemCount: number
  skippedItems: Array<{ name: string; type: string }>
}

export const useZabbixTemplatesStore = defineStore('zabbixTemplates', () => {
  const templates = ref<ZabbixTemplate[]>([])
  const loading = ref(false)
  const importing = ref(false)
  const error = ref<string | null>(null)

  async function fetchTemplates() {
    loading.value = true
    error.value = null
    try {
      templates.value = await apiService.get<ZabbixTemplate[]>('/zabbix-templates')
    } catch (err: unknown) {
      error.value = err instanceof Error ? err.message : 'Erro ao carregar templates Zabbix'
    } finally {
      loading.value = false
    }
  }

  async function importTemplate(content: string): Promise<ZabbixImportResult[] | null> {
    importing.value = true
    error.value = null
    try {
      const result = await apiService.post<{ templates: ZabbixImportResult[] }>(
        '/zabbix-templates',
        {
          content,
        }
      )
      await fetchTemplates()
      return result.templates
    } catch (err: unknown) {
      error.value = err instanceof Error ? err.message : 'Erro ao importar template Zabbix'
      return null
    } finally {
      importing.value = false
    }
  }

  async function deleteTemplate(id: number): Promise<boolean> {
    try {
      await apiService.delete(`/zabbix-templates/${id}`)
      templates.value = templates.value.filter((t) => t.id !== id)
      return true
    } catch (err: unknown) {
      error.value = err instanceof Error ? err.message : 'Erro ao excluir template Zabbix'
      return false
    }
  }

  return {
    templates,
    loading,
    importing,
    error,
    fetchTemplates,
    importTemplate,
    deleteTemplate,
  }
})

File: backend-rust/src/controllers/zabbix_templates.rs
//! Importacao e gerenciamento de templates Zabbix.

use axum::{http::StatusCode, response::IntoResponse};
use chrono::Utc;
use loco_rs::prelude::*;
use sea_orm::{
    ColumnTrait, EntityTrait, PaginatorTrait, QueryFilter, QueryOrder, Set, TransactionTrait,
};

use crate::{
    dtos::resources::ZabbixImportInput,
    models::_entities::{devices, zabbix_template_items, zabbix_templates},
    services::{
        maintenance::resource_cleanup::ResourceCleanupService,
        shared::errors::{AppError, AppResult},
        zabbix::parser,
    },
};

async fn index(State(ctx): State<AppContext>) -> AppResult<Response> {
    let templates = zabbix_templates::Entity::find()
        .order_by_asc(zabbix_templates::Column::Name)
        .all(&ctx.db)
        .await?;
    let mut output = Vec::new();
    for template in templates {
        let items = zabbix_template_items::Entity::find()
            .filter(zabbix_template_items::Column::TemplateId.eq(template.id))
            .all(&ctx.db)
            .await?;
        let count = devices::Entity::find()
            .filter(devices::Column::ZabbixTemplateId.eq(template.id))
            .count(&ctx.db)
            .await?;
        output.push(serde_json::json!({
            "id": template.id,
            "zabbixUuid": template.zabbix_uuid,
            "name": template.name,
            "description": template.description,
            "zabbixVersion": template.zabbix_version,
            "importedAt": template.imported_at.to_rfc3339(),
            "deviceCount": count,
            "items": items.into_iter().map(|item| serde_json::json!({
                "id": item.id,
                "name": item.name,
                "key": item.key,
                "snmpOid": item.snmp_oid,
                "valueType": item.value_type,
                "units": item.units,
                "multiplier": item.multiplier,
            })).collect::<Vec<_>>(),
        }));
    }
    Ok(format::json(output)?)
}

async fn show(State(ctx): State<AppContext>, Path(id): Path<i64>) -> AppResult<Response> {
    let template = zabbix_templates::Entity::find_by_id(id)
        .one(&ctx.db)
        .await?
        .ok_or_else(|| AppError::not_found("Template Zabbix não encontrado"))?;
    let items = zabbix_template_items::Entity::find()
        .filter(zabbix_template_items::Column::TemplateId.eq(id))
        .all(&ctx.db)
        .await?;
    Ok(format::json(serde_json::json!({
        "id": template.id,
        "zabbixUuid": template.zabbix_uuid,
        "name": template.name,
        "description": template.description,
        "zabbixVersion": template.zabbix_version,
        "rawExport": template.raw_export,
        "importedAt": template.imported_at.to_rfc3339(),
        "items": items,
    }))?)
}

async fn store(
    State(ctx): State<AppContext>,
    Json(input): Json<ZabbixImportInput>,
) -> AppResult<Response> {
    let parsed = parser::parse_zabbix_template_export(&input.content)
        .map_err(|error| AppError::validation(error.to_string()))?;
    let txn = ctx.db.begin().await?;
    let mut imported = Vec::new();
    for template in parsed.templates {
        let existing = match &template.uuid {
            Some(uuid) => {
                zabbix_templates::Entity::find()
                    .filter(zabbix_templates::Column::ZabbixUuid.eq(uuid))
                    .one(&txn)
                    .await?
            }
            None => None,
        };
        let row = if let Some(existing) = existing {
            zabbix_template_items::Entity::delete_many()
                .filter(zabbix_template_items::Column::TemplateId.eq(existing.id))
                .exec(&txn)
                .await?;
            let mut active: zabbix_templates::ActiveModel = existing.into();
            active.name = Set(template.name.clone());
            active.description = Set(template.description.clone());
            active.zabbix_version = Set(template.version.clone());
            active.raw_export = Set(parsed.raw_export.clone());
            active.imported_at = Set(Utc::now().into());
            active.update(&txn).await?
        } else {
            zabbix_templates::ActiveModel {
                zabbix_uuid: Set(template.uuid.clone()),
                name: Set(template.name.clone()),
                description: Set(template.description.clone()),
                zabbix_version: Set(template.version.clone()),
                raw_export: Set(parsed.raw_export.clone()),
                imported_at: Set(Utc::now().into()),
                ..Default::default()
            }
            .insert(&txn)
            .await?
        };
        let item_count = template.items.len();
        for item in template.items {
            zabbix_template_items::ActiveModel {
                template_id: Set(row.id),
                zabbix_uuid: Set(item.uuid),
                name: Set(item.name),
                key: Set(item.key),
                snmp_oid: Set(item.snmp_oid),
                value_type: Set(item.value_type),
                units: Set(item.units),
                multiplier: Set(item.multiplier),
                ..Default::default()
            }
            .insert(&txn)
            .await?;
        }
        imported.push(serde_json::json!({
            "id": row.id,
            "name": row.name,
            "itemCount": item_count,
            "skippedItems": template.skipped_items,
        }));
    }
    txn.commit().await?;
    Ok((
        StatusCode::CREATED,
        Json(serde_json::json!({ "templates": imported })),
    )
        .into_response())
}

async fn destroy(State(ctx): State<AppContext>, Path(id): Path<i64>) -> AppResult<Response> {
    zabbix_templates::Entity::find_by_id(id)
        .one(&ctx.db)
        .await?
        .ok_or_else(|| AppError::not_found("Template Zabbix não encontrado"))?;
    ResourceCleanupService::delete_zabbix_template(&ctx.db, id).await?;
    Ok(StatusCode::NO_CONTENT.into_response())
}

pub fn routes() -> Routes {
    Routes::new()
        .prefix("/zabbix-templates")
        .add("/", get(index).post(store))
        .add("/{id}", get(show).delete(destroy))
}
